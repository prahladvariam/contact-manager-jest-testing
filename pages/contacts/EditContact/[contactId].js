import { useState } from "react";
import { useRouter } from "next/router";
import "bootstrap/dist/css/bootstrap.min.css";
import Link from "next/link";
import React from "react";

function EditContact({ contact }) {
  const router = useRouter();
  const state = contact;
  const [newContact, setContact] = useState();
  const contactId = router.query.contactId;
  //console.log(newContact);
  const submitContact = async (e) => {
    e.preventDefault();
    const response = await fetch(
      `http://localhost:4000/contacts/${contact.id}`,
      {
        method: "PUT",
        body: JSON.stringify(newContact),
        headers: {
          "Content-Type": "application/json",
        },
      }
    );
    // console.log(newContact);
    const data = await response.json();
    if (data !== "") {
      alert("Data saved successfully!");
    }
    router.push("/contacts");
  };

  return (
    <div className="container-fluid">
      <div className="row">
        <h2 className="text-center bg-dark text-white">CONTACT MANAGER</h2>
      </div>
      <div className="container">
        <div className="row" style={{ padding: "35px" }}>
          <div className="col-md-3"></div>
          <div
            className="col-md-6"
            style={{
              backgroundColor: "#012641",
              borderRadius: "10px 10px 10px 10px",
              color: "#ffffff",
              padding: "40px",
            }}
            key={contact.id}
          >
            <h4>
              <form onSubmit={submitContact}>
                <input type="text" defaultValue={state.id} hidden />
                <label>Name:</label>
                <input
                  type="text"
                  className="form-control"
                  defaultValue={contact.name}
                  onChange={(e) => {
                    state.name = e.target.value;
                  }}
                  data-testid="name"
                />
                <label>Mobile:</label>
                <input
                  type="text"
                  className="form-control"
                  defaultValue={contact.mobile}
                  onChange={(e) => {
                    state.mobile = e.target.value;
                  }}
                  data-testid="number"
                />
                <label>Email:</label>
                <input
                  type="text"
                  className="form-control"
                  defaultValue={contact.email}
                  onChange={(e) => {
                    state.email = e.target.value;
                  }}
                  data-testid="email"
                />
                <div className="text-center">
                  <button
                    className="btn btn-primary m-2"
                    onClick={(e) => setContact(state)}
                  >
                    Save
                  </button>
                </div>
              </form>
            </h4>
          </div>
          <div className="col-md-3"></div>
        </div>
        <div className="row">
          <div className="text-center">
            <Link href="/contacts" className="text-center">
              <button className="btn btn-primary m-4">
                Back To Contact List
              </button>
            </Link>
          </div>
        </div>
      </div>
    </div>
  );
}

export default EditContact;

export async function getStaticPaths() {
  const response = await fetch("http://localhost:4000/contacts");
  const contacts = await response.json();
  const paths = contacts.map((contact) => ({
    params: { contactId: contact.id.toString() },
  }));

  return {
    paths,
    fallback: false,
  };
}

export async function getStaticProps(context) {
  console.log("Regenrating Data");
  const { params } = context;
  const response = await fetch(
    `http://localhost:4000/contacts/${params.contactId}`
  );
  const data = await response.json();

  return {
    props: {
      contact: data,
    },
  };
}
