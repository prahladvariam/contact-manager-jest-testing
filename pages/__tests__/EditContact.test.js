import React from 'react';
import EditContact from '../contacts/EditContact/[contactId]';
import singletonRouter from "next/router";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import mockRouter from "next-router-mock";

jest.mock("next/router", () => require("next-router-mock"));
jest.mock("next/dist/client/router", () => require("next-router-mock"));
describe("Edit Contact Tests",() => {
    const contact = 
        {
            id: 1,
            name: "Sample1",
            email: "sample1@mail.com",
            mobile: 5451651845
        }
  const jestTest = jest.fn()

  const unmockedFetch = global.fetch

    beforeAll(() => {
        jest.setTimeout(90 * 1000)
        global.fetch = () =>
        Promise.resolve({
        json: () => Promise.resolve([]),
        })
    })

    afterAll(() => {
        global.fetch = unmockedFetch
    })


  beforeEach(() => {
    jest.resetModules();
    mockRouter.setCurrentUrl("/intials");
  });

  it("Add contact page is rendering correctly", () => {
    const component = render(<EditContact contact={contact} />)
    expect(component).toMatchSnapshot()
  })


  it("onSubmit is Called", async() => {
    const jsdomAlert = window.alert;
    window.alert = () => {};  
    render(<EditContact contact={contact} />)
    userEvent.type(screen.getByTestId('name'),"new value")
    userEvent.type(screen.getByTestId('number'), "12345")
    userEvent.type(screen.getByTestId('email'),"new@mail.com")
    userEvent.click(screen.getByRole('button', {
      name: /save/i
    }))
    await waitFor(() => {
      expect(singletonRouter).toMatchObject({asPath: "/contacts"})
    })
    window.alert = jsdomAlert; 
  })

})
