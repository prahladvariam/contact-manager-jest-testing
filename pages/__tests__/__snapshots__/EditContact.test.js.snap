// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Contact Tests Add contact page is rendering correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container-fluid"
      >
        <div
          class="row"
        >
          <h2
            class="text-center bg-dark text-white"
          >
            CONTACT MANAGER
          </h2>
        </div>
        <div
          class="container"
        >
          <div
            class="row"
            style="padding: 35px;"
          >
            <div
              class="col-md-3"
            />
            <div
              class="col-md-6"
              style="background-color: rgb(1, 38, 65); border-radius: 10px 10px 10px 10px; color: rgb(255, 255, 255); padding: 40px;"
            >
              <h4>
                <form>
                  <input
                    hidden=""
                    type="text"
                    value="1"
                  />
                  <label>
                    Name:
                  </label>
                  <input
                    class="form-control"
                    data-testid="name"
                    type="text"
                    value="Sample1"
                  />
                  <label>
                    Mobile:
                  </label>
                  <input
                    class="form-control"
                    data-testid="number"
                    type="text"
                    value="5451651845"
                  />
                  <label>
                    Email:
                  </label>
                  <input
                    class="form-control"
                    data-testid="email"
                    type="text"
                    value="sample1@mail.com"
                  />
                  <div
                    class="text-center"
                  >
                    <button
                      class="btn btn-primary m-2"
                    >
                      Save
                    </button>
                  </div>
                </form>
              </h4>
            </div>
            <div
              class="col-md-3"
            />
          </div>
          <div
            class="row"
          >
            <div
              class="text-center"
            >
              <button
                class="btn btn-primary m-4"
              >
                Back To Contact List
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container-fluid"
    >
      <div
        class="row"
      >
        <h2
          class="text-center bg-dark text-white"
        >
          CONTACT MANAGER
        </h2>
      </div>
      <div
        class="container"
      >
        <div
          class="row"
          style="padding: 35px;"
        >
          <div
            class="col-md-3"
          />
          <div
            class="col-md-6"
            style="background-color: rgb(1, 38, 65); border-radius: 10px 10px 10px 10px; color: rgb(255, 255, 255); padding: 40px;"
          >
            <h4>
              <form>
                <input
                  hidden=""
                  type="text"
                  value="1"
                />
                <label>
                  Name:
                </label>
                <input
                  class="form-control"
                  data-testid="name"
                  type="text"
                  value="Sample1"
                />
                <label>
                  Mobile:
                </label>
                <input
                  class="form-control"
                  data-testid="number"
                  type="text"
                  value="5451651845"
                />
                <label>
                  Email:
                </label>
                <input
                  class="form-control"
                  data-testid="email"
                  type="text"
                  value="sample1@mail.com"
                />
                <div
                  class="text-center"
                >
                  <button
                    class="btn btn-primary m-2"
                  >
                    Save
                  </button>
                </div>
              </form>
            </h4>
          </div>
          <div
            class="col-md-3"
          />
        </div>
        <div
          class="row"
        >
          <div
            class="text-center"
          >
            <button
              class="btn btn-primary m-4"
            >
              Back To Contact List
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
